Add docs for SSZ Query package (#16299)
prysmaticlabs/prysm: encoding/ssz/query/doc.md

# SSZ Query Package

The `encoding/ssz/query` package provides a system for analyzing, querying, and generating Merkle proofs for SSZ ([Simple Serialize](https://github.com/ethereum/consensus-specs/blob/master/ssz/simple-serialize.md)) data structures. It enables runtime analysis of SSZ-serialized Go objects with reflection, path-based queries through nested structures, generalized index calculation, and Merkle proof generation.

> This package is designed to be **generic**, which means it can analyze/prove any SSZ objects including `BeaconState` and `BeaconBlock`.

Reviewer: ```suggestion
This package is designed to be **generic**, it operates on arbitrary SSZ-serialized Go values at runtime, so the same query/proof machinery applies equally to core consensus types like BeaconState/BeaconBlock and to any customized SSZ schemas.
```

// 4. Generate a Merkle proof
proof, err := info.Prove(gindex)

// 5. Get offset and length to slice the SSZ-encoded bytes
sszBytes, _ := block.MarshalSSZ()
_, offset, length, err := query.CalculateOffsetAndLength(info, path)
// slotBytes contains the SSZ-encoded value at the queried path
slotBytes := sszBytes[offset : offset+length]
```

## Exported API

The main exported API consists of:

```go
// AnalyzeObject analyzes an SSZ object and returns its structural information
func AnalyzeObject(obj SSZObject) (*SszInfo, error)

// ParsePath parses a path string like ".field1.field2[0].field3"
func ParsePath(rawPath string) (Path, error)

// CalculateOffsetAndLength computes byte offset and length for a path within an SSZ object
func CalculateOffsetAndLength(sszInfo *SszInfo, path Path) (*SszInfo, uint64, uint64, error)

// GetGeneralizedIndexFromPath calculates the generalized index for a given path
func GetGeneralizedIndexFromPath(info *SszInfo, path Path) (uint64, error)

// Prove generates a Merkle proof for a target generalized index
func (s *SszInfo) Prove(gindex uint64) (*fastssz.Proof, error)
```

## Type System

### SSZ Types

> The package now supports [all standard SSZ types](https://github.com/ethereum/consensus-specs/blob/master/ssz/simple-serialize.md#typing) except `ProgressiveList` and `Union`.

Reviewer: ```suggestion
The package now supports [all standard SSZ types](https://github.com/ethereum/consensus-specs/blob/master/ssz/simple-serialize.md#typing) except `ProgressiveList`, `ProgressiveContainer`, `ProgressiveBitlist`, `Union` and `CompatibleUnion`.
```

### Path Parsing (`path.go`)

The `ParsePath` function parses path strings with the following rules:

- **Dot notation**: `.field1.field2` for field access
- **Array indexing**: `[0]`, `[42]` for element access
- **Length queries**: `len(.field)` for list/vector lengths
- **Character set**: Only `[A-Za-z0-9._\[\]\(\)]` allowed

Example:
```go
path, _ := ParsePath(".nested.array_field[5].inner_field")
// Returns: Path{
//   Elements: [
//     PathElement{Name: "nested"},
//     PathElement{Name: "array_field", Index: <Pointer to uint64(5)>},
//     PathElement{Name: "inner_field"}
//   ]
// }
```

### Generalized Index Calculation (`generalized_index.go`)

The generalized index is a tree position identifier. This package follows the [Ethereum consensus-specs](https://github.com/ethereum/consensus-specs/blob/master/ssz/merkle-proofs.md#generalized-merkle-tree-index) to calculate the generalized index.

### Merkle Proof Generation (`merkle_proof.go`, `proof_collector.go`)

The `Prove` method generates Merkle proofs using a single-sweep merkleization algorithm:

#### Algorithm Overview

1. **Registration Phase** (`addTarget`)
   - Marks target gindex and all required sibling nodes

>    - Walks from target leaf to root (gindex=1)

Reviewer: ```suggestion
   - Marks the target gindex
   - Marks every required sibling node by walking from the target leaf to the root (gindex=1).
```

2. **Merkleization Phase** (`merkleize`)
   - Traverses data structure **recursively**
   - Builds Merkle tree from leaves to root
   - Collects hashes at registered gindices during traversal

3. **Proof Conversion Phase** (`toProof`)
   - Converts collected hashes to `fastssz.Proof`
   - Stores leaf value at proof index
   - Orders sibling hashes from leaf upward

#### Type-Specific Merkleization

- **Basic Types**: Serializes to 32-byte leaf, little-endian, zero-padded
- **Containers**: Merkleizes each field, then merkleizes field roots
- **Vectors**: Fixed-size collections with direct merkleization
- **Lists**: Two-part merkleization (data subtree + length mixin)

> - **Bitlists/Bitvectors**: Packs bits into 32-byte chunks

Reviewer: ```suggestion
   - **Basic Types**: Merkleize as a single 32-byte chunk containing the SSZ-encoded value, little-endian and zero-padded.
   - **Containers**: Merkleize by computing the 32-byte root of each field, then merkleizing the field roots.
   - **Vectors**: Merkleize fixed-size collections by chunking elements (basic types are packed; composite types use their 32-byte roots).
   - **Lists**: Merkleize variable-size collections by chunking elements up to the type’s maximum, then mix in the actual element count.
   - **Bitvectors**: Merkleize packed bits in 32-byte, LSB-first chunks, using the type’s chunk count.
   - **Bitlists**: Merkleize packed bits in 32-byte, LSB-first chunks (excluding the termination bit), using the type’s chunk count, then mix in the bit length.
```

# SSZ Query Package

> The `encoding/ssz/query` package provides a system for analyzing, querying, and generating Merkle proofs for SSZ ([Simple Serialize](https://github.com/ethereum/consensus-specs/blob/master/ssz/simple-serialize.md)) data structures. It enables runtime analysis of SSZ-serialized Go objects with reflection, path-based queries through nested structures, generalized index calculation, and Merkle proof generation.

Reviewer 2: > for analyzing, querying, and generating Merkle proofs

this sounds as if we analyze proofs

Author: Rephrased the sentence to be more clearer.

> This package is designed to be **generic**, it operates on arbitrary SSZ-serialized Go values at runtime, so the same query/proof machinery applies equally to core consensus types like BeaconState/BeaconBlock and to any customized SSZ schemas.

Reviewer 2: ```suggestion
This package is designed to be generic. It operates on arbitrary SSZ-serialized Go values at runtime, so the same query/proof machinery applies equally to any SSZ type, including the BeaconState/BeaconBlock.
```

AIs like to emphasize certain words, but to me it reads weirdly. Also `customized SSZ schemas` sounds as if we're talking about some non-standard SSZ schema

// Prove generates a Merkle proof for a target generalized index
func (s *SszInfo) Prove(gindex uint64) (*fastssz.Proof, error)
```

## Type System

### SSZ Types

The package now supports [all standard SSZ types](https://github.com/ethereum/consensus-specs/blob/master/ssz/simple-serialize.md#typing) except `ProgressiveList`, `ProgressiveContainer`, `ProgressiveBitlist`, `Union` and `CompatibleUnion`.

### Core Data Structures

#### `SszInfo`

The `SszInfo` structure contains complete structural metadata for an SSZ type:

```go
type SszInfo struct {
   sszType       SszType           // SSZ Type classification
   typ           reflect.Type      // Go reflect.Type
   source        SSZObject         // Original SSZObject reference. Mostly used for reusing SSZ methods like `HashTreeRoot`.
   isVariable    bool              // True if contains variable-size fields

   // Composite types have corresponding metadata. Other fields would be nil except for the current type.
   containerInfo *containerInfo
   listInfo      *listInfo
   vectorInfo    *vectorInfo
   bitlistInfo   *bitlistInfo
   bitvectorInfo *bitvectorInfo
}
```

#### `Path`

> The `Path` structure represents navigation paths through SSZ structures. It supports accessing a field by field name, accessing an element by index (list/vector type), and length of homogenous collection types. `ParsePath` function parses a raw string into `Path` instance, which is commonly used in other APIs like `CalculateOffsetAndLength` and `GetGeneralizedIndexFromPath`.

Reviewer 2: ```suggestion
The `Path` structure represents navigation paths through SSZ structures. It supports accessing a field by field name, accessing an element by index (list/vector type), and finding the length of homogenous collection types. The `ParsePath` function parses a raw string into a `Path` instance, which is commonly used in other APIs like `CalculateOffsetAndLength` and `GetGeneralizedIndexFromPath`.
```

```go
type Path struct {
   Length   bool           // Flag for length queries (e.g., len(.field))
   Elements []PathElement  // Sequence of field accesses and indices
}

type PathElement struct {
   Name  string  // Field name
   Index *uint64 // list/vector index (nil if not an index access)
}
```

## Implementation Details

### Type Analysis (`analyzer.go`)

> The `AnalyzeObject` function performs **recursive** type introspection using Go reflection:

Reviewer 2: ```suggestion
The `AnalyzeObject` function performs recursive type introspection using Go reflection:
```
the fact that it's recursive isn't so important

1. **Type Inspection** - Examines Go `reflect.Value` to determine SSZ type

>    - Basic types: `uint8`, `uint16`, `uint32`, `uint64`, `bool` → SSZType constants

Reviewer 2: ```suggestion
   - Basic types (`uint8`, `uint16`, `uint32`, `uint64`, `bool`): `SSZType` constants
```

>    - Slices: Determined from struct tags. (`ssz-size` for vectors, `ssz-max` for lists) There is a related [write-up](https://hackmd.io/@junsong/H101DKnwxl) regarding struct tags.

Reviewer 2: ```suggestion
   - Slices: Determined from struct tags (`ssz-size` for vectors, `ssz-max` for lists). There is a related [write-up](https://hackmd.io/@junsong/H101DKnwxl) regarding struct tags.
```

- **Array indexing**: `[0]`, `[42]` for element access
- **Length queries**: `len(.field)` for list/vector lengths
- **Character set**: Only `[A-Za-z0-9._\[\]\(\)]` allowed

Example:
```go
path, _ := ParsePath(".nested.array_field[5].inner_field")
// Returns: Path{
//   Elements: [
//     PathElement{Name: "nested"},
//     PathElement{Name: "array_field", Index: <Pointer to uint64(5)>},
//     PathElement{Name: "inner_field"}
//   ]
// }
```

### Generalized Index Calculation (`generalized_index.go`)

The generalized index is a tree position identifier. This package follows the [Ethereum consensus-specs](https://github.com/ethereum/consensus-specs/blob/master/ssz/merkle-proofs.md#generalized-merkle-tree-index) to calculate the generalized index.

### Merkle Proof Generation (`merkle_proof.go`, `proof_collector.go`)

The `Prove` method generates Merkle proofs using a single-sweep merkleization algorithm:

#### Algorithm Overview

1. **Registration Phase** (`addTarget`)
   - Marks the target gindex
   - Marks every required sibling node by walking from the target leaf to the root (gindex=1).

2. **Merkleization Phase** (`merkleize`)

>    - Traverses data structure **recursively**

Reviewer 2: ```suggestion
   - Traverses data structure recursively
```

   - Builds Merkle tree from leaves to root
   - Collects hashes at registered gindices during traversal

> 3. **Proof Conversion Phase** (`toProof`)

Reviewer 2: I had to look at the code to see how this works, which means that this explanation is probably too cryptic

Author: Actually `toProof` is more like "reassembling" the proof into compatible format. I fixed this a bit but let me know what you think! Maybe we might exclude this part.

   - Converts collected hashes to `fastssz.Proof`
   - Stores leaf value at proof index
   - Orders sibling hashes from leaf upward

> #### Type-Specific Merkleization

Reviewer 2: That's just standard SSZ merkleization algorithm, there's no need to include it in the doc

   - **Basic Types**: Merkleize as a single 32-byte chunk containing the SSZ-encoded value, little-endian and zero-padded.
   - **Containers**: Merkleize by computing the 32-byte root of each field, then merkleizing the field roots.
   - **Vectors**: Merkleize fixed-size collections by chunking elements (basic types are packed; composite types use their 32-byte roots).
   - **Lists**: Merkleize variable-size collections by chunking elements up to the type’s maximum, then mix in the actual element count.
   - **Bitvectors**: Merkleize packed bits in 32-byte, LSB-first chunks, using the type’s chunk count.
   - **Bitlists**: Merkleize packed bits in 32-byte, LSB-first chunks (excluding the termination bit), using the type’s chunk count, then mix in the bit length.

> ## Dependencies

Reviewer 2: This section is not needed. It's not explaining how the package works, it can quickly become obsolete and this information can be easily found by inspecting bazel files or through the use of some tooling

//     PathElement{Name: "nested"},
//     PathElement{Name: "array_field", Index: <Pointer to uint64(5)>},
//     PathElement{Name: "inner_field"}
//   ]
// }
```

### Generalized Index Calculation (`generalized_index.go`)

The generalized index is a tree position identifier. This package follows the [Ethereum consensus-specs](https://github.com/ethereum/consensus-specs/blob/master/ssz/merkle-proofs.md#generalized-merkle-tree-index) to calculate the generalized index.

### Merkle Proof Generation (`merkle_proof.go`, `proof_collector.go`)

The `Prove` method generates Merkle proofs using a single-sweep merkleization algorithm:

#### Algorithm Overview

**Key Terms:**

- **Target gindex** (generalized index): The position of the SSZ element you want to prove, expressed as a generalized Merkle tree index.
  - Note: The generalized index for root is 1.
- **Registered gindices**: The set of tree positions whose node hashes must be captured during merkleization in order to later assemble the proof.
- **Sibling node**: The node that shares the same parent as another node.
- **Leaf value**: The 32-byte hash of the target node (the node being proven). Stored in `Proof.Leaf`.

> - **Proof index**: The generalized index of the target node. Stored in `Proof.Index`.

Reviewer 2: This seems to be exactly the same as the `Target gindex` above - can we use just one term?

Author: Consolidated two items into one!
